### PR TITLE
Implementing Rest API

### PR DESCRIPTION
**Pull request by mdumbrava20**

### CODE REVIEW

File: src/main/java/com/endava/webapp/controllers/DepartmentController.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+import javax.validation.Valid;
+import java.util.List;
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/departments")
+public class DepartmentController {
```
Comment (vermurachi) [inline src/main/java/com/endava/webapp/controllers/DepartmentController.java]: Create interfaces for controller classes

File: src/main/java/com/endava/webapp/models/dto/DepartmentDTO.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
+package com.endava.webapp.models.dto;
```
Comment (vermurachi) [inline src/main/java/com/endava/webapp/models/dto/DepartmentDTO.java]: move dto package to the domain package

File: src/main/java/com/endava/webapp/models/Employee.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
+package com.endava.webapp.models;
```
Comment (vermurachi) [inline src/main/java/com/endava/webapp/models/Employee.java]: move model package to domain

File: src/main/java/com/endava/webapp/domain/service/DepartmentService.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package com.endava.webapp.domain.service;
+
```
Comment (vermurachi) [inline src/main/java/com/endava/webapp/domain/service/DepartmentService.java]: Create a package department in service package and add DepartmentService, *Impl and DepartmentNotFoundException.
Same for employee.

File: src/main/java/com/endava/webapp/controllers/DepartmentController.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+    private final DepartmentService departmentService;
+
+    @GetMapping()
+    public List<Department> getAllDepartments(){
```
Comment (aledimov) [inline src/main/java/com/endava/webapp/controllers/DepartmentController.java]: Make consistent with others. Return ResponseEntity here also

File: src/main/java/com/endava/webapp/controllers/DepartmentController.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+    @PostMapping()
+    public ResponseEntity<Department> addDepartment(@RequestBody @Valid DepartmentDTO department){
+        Department savedDepartment = departmentService.saveDepartment(department.toDepartment());
+        return ResponseEntity.status(HttpStatus.CREATED).body(savedDepartment);
+    }
+
+    @PutMapping("/{id}")
+    public ResponseEntity<Department> editDepartment(@RequestBody @Valid DepartmentDTO department, @PathVariable long id){
+        Department oldDepartment = departmentService.getById(id);
```
Comment (aledimov) [inline src/main/java/com/endava/webapp/controllers/DepartmentController.java]: Can we perform this operation in service layer?

File: src/main/java/com/endava/webapp/models/dto/EmployeeDTO.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+import javax.validation.constraints.NotNull;
+
+@Setter
+@Getter
+@AllArgsConstructor
+@NoArgsConstructor
+public class EmployeeDTO {
+
+    @NotNull
+    private long departmentId;
```
Comment (aledimov) [inline src/main/java/com/endava/webapp/models/dto/EmployeeDTO.java]: Will it work on primitives? If you'll not add a value for ID will it throw an exception or assign a default value for primitive type?

File: src/main/java/com/endava/webapp/models/dto/EmployeeDTO.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    @NotEmpty
+    @NotBlank
+    private String email;
+
+    @NotNull
+    @NotEmpty
+    @NotBlank
+    private String phoneNumber;
+
+    @NotNull
```
Comment (aledimov) [inline src/main/java/com/endava/webapp/models/dto/EmployeeDTO.java]: Will it work on primitives?

File: src/main/java/com/endava/webapp/models/dto/EmployeeDTO.java [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    private String firstName;
+
+    @NotNull
+    @NotEmpty
+    @NotBlank
+    private String lastName;
+
+    @NotNull
+    @NotEmpty
+    @NotBlank
```
Comment (aledimov) [inline src/main/java/com/endava/webapp/models/dto/EmployeeDTO.java]: NotBlank = NotEmpy + NotNull?

File: src/test/java/com/endava/webapp/service/DepartmentServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+
+class DepartmentServiceTest {
+
+    private final DepartmentRepository departmentRepository = mock(DepartmentRepository.class);
```
Comment (aledimov) [inline src/test/java/com/endava/webapp/service/DepartmentServiceTest.java:27]: Make consistent. Use MockBean

File: src/main/java/com/endava/webapp/domain/service/DepartmentService.java (hunk 4 above)
Comment (mdumbrava20) [reply src/main/java/com/endava/webapp/domain/service/DepartmentService.java]: Fixed.

File: src/main/java/com/endava/webapp/models/Employee.java (hunk 3 above)
Comment (mdumbrava20) [reply src/main/java/com/endava/webapp/models/Employee.java]: Fixed.

File: src/main/java/com/endava/webapp/models/dto/DepartmentDTO.java (hunk 2 above)
Comment (mdumbrava20) [reply src/main/java/com/endava/webapp/models/dto/DepartmentDTO.java]: Fixed.

File: src/main/java/com/endava/webapp/controllers/DepartmentController.java (hunk 1 above)
Comment (mdumbrava20) [reply src/main/java/com/endava/webapp/controllers/DepartmentController.java]: Fixed.

File: src/main/java/com/endava/webapp/models/dto/EmployeeDTO.java (hunk 9 above)
Comment (mdumbrava20) [reply src/main/java/com/endava/webapp/models/dto/EmployeeDTO.java]: Fixed. NotBlank means at least one character so it includes (NotEmpty and NotNull).

File: src/main/java/com/endava/webapp/controllers/DepartmentController.java (hunk 5 above)
Comment (mdumbrava20) [reply src/main/java/com/endava/webapp/controllers/DepartmentController.java]: Fixed.

File: src/test/java/com/endava/webapp/service/DepartmentServiceTest.java (hunk 10 above)
Comment (mdumbrava20) [reply src/test/java/com/endava/webapp/service/DepartmentServiceTest.java:27]: Fixed.

File: src/main/java/com/endava/webapp/models/dto/EmployeeDTO.java (hunk 8 above)
Comment (mdumbrava20) [reply src/main/java/com/endava/webapp/models/dto/EmployeeDTO.java]: Solved.

File: src/main/java/com/endava/webapp/models/dto/EmployeeDTO.java (hunk 7 above)
Comment (mdumbrava20) [reply src/main/java/com/endava/webapp/models/dto/EmployeeDTO.java]: Solved.

File: src/main/java/com/endava/webapp/controllers/DepartmentController.java (hunk 6 above)
Comment (mdumbrava20) [reply src/main/java/com/endava/webapp/controllers/DepartmentController.java]: Solved.

**Review by vermurachi — APPROVED** [review]

**Review by aledimov — APPROVED** [review]
LGTM

**Pull request merged**